documentation/design-idna.md

```diff
@@ -0,0 +1,244 @@
+# DNSControl and Internationalized domain name
+
+This is my proposal for how to make IDNs work better in DNSControl.
+Basically, the UI will accept any format.  Early in the process
+DNSControl will store labels/domains data 4 ways: As received from the
+user (downcased), ASCII, Unicode, and in a "display" format that shows
+both. The conversions already done ahead of time, providers can
+access whatever format they need.  Output from the main program will
+use the "display" format when possible.
+
+# Problem Statement
+
+DNSControl doesn't handle internationalized domain names (IDNs) very
+well.  Coverage is uneven: They work better in some providers than
+others. There are bugs and inconsistencies. Writing a provider that
+handles IDNs properly requires doing most of the work in the provider
+itself, which means every provider maintainer must be an expert in
+IDNs, which is unreasonable.
+
+# Background:
+
+RFC 3490 recommends how applications should handle IDNs. My summary:
+(1) the UI should accept a mix of Unicode and ASCII domains/labels.
+(2) internally translate everything to ASCII (punycode) and do all
+processing in that format, (3) when displaying output, display it as
+the user input it, or Unicode, or ASCII, or give users a choice.
+
+* IDNA: Internationalizing Domain Names in Applications
+* IDNs: Internationalized domain names
+* ACE Prefix: The `xn--` that means "Puny code follows"
+* ASCII: A label or domain is output as ASCII with ACE prefix if needed.
+* Unicode: A label or domain is output as Unicode.
+
+Proposed Outcome:
+
+1. Users should be able to input domains and labels in either ASCII (with ACE prefix if needed), Unicode, or a mix.  This holds for input via `dnsconfig.js` (domain names, labels, and targets); as well as flags such as `--domains`.
+2a. Output should be Unicode or both with the ASCII being in parenthesis. Example: `рф.com (xn--p1ai.com)`
+2b. Or maybe the reverse? Example: `xn--p1ai.com (рф.com)`
+3. DNSControl's main code should create a "paved path" for providers to make it easy for them to do the right thing. It should be easier to do the right thing than the wrong thing.  The default (i.e. "lazy") path should result in the behavior we desire.
+
+Here are some example outputs:
+
+NOTE: Feedback needed!  Do you prefer "a" or "b"?  Is there an even better format I should consider?  Should we use `{}` instead of `()`?
```

Comment:
I prefer `b` and `()`. We should also specify `NameDisplay` only is a string with both variants if the domain is non-ASCII.